### DISCUSSION TITLE
Offline support similar to Draw.io / Diagrams.net / Nomie

### DISCUSSION THREAD

**Discussion by sathyamvellal** [TriliumNext/Trilium#3752] [Ideas]
Draw.io / Diagrams.net installs as a PWA on mobile / tablet devices and also have offline support. I haven't personally looked into how these apps achieve that functionality, but I wanted to ask @zadam, and perhaps others who have tinkered around with Trilium code, for feasibility in achieving this. If its possible to utilize browser local storage (or maybe another/better API if there is one), I'd like the community's help for tips and tinkering around with this. 

To the limited extent I've looked into, Nomie uses PouchDB to store data locally (and offline), and syncs data remotely whenever connection is available.

If this is indeed feasible, then I'd like to volunteer to bring offline support to Trilium, no matter the effort / timeline involved, as this would be a game-changer for me in my day-to-day workflow. Appreciate your help and support! Having this would quite possibility eliminate the need for having full-featured native apps (assuming only reason for a native app is offline support).

Source for diagrams.net: https://github.com/jgraph/drawio
Source for Nomie app: https://github.com/open-nomie/nomie6-oss

**Comment by XXXJetfireXXX** [comment]
I might have misunderstood you, but afaik Trilium already supports offline usage on Linux and Windows systems (I'm using the local setup myself). If you're talking about a mobile app specifically though...

**Comment by sathyamvellal** [comment]
Ah I can see how my description can be confusing. My apologies. (I'll edit the description once my clarification in the comments make sense). On a desktop/laptop, I can have Trilium work locally, yes, as the system hosts the instance.

An example of what I intended to describe is installing a PWA of a server-hosted Trilium instance on an iPad. Today, this PWA is directly loading the remote instance. Without internet connection, it is not able to render anything. 
Apps like diagrams.net and nomie are capable of rendering pages / components offline, have the data stored in perhaps browser local storage, and sync with a remote server whenever internet is available. I haven't read about / code-read how these apps achieve that, so I'm not sure exactly how yet.

In my mind, I'm thinking - Trilium installed as a PWA is self-contained in that, it doesn't need a remote server to send data via api (since data would be on browser local storage), and my guess the PWA already has the frontend packaged, so is able to render a remote-hosted trilium instance offline (since remote-api calls aren't required), while being able to sync data when possible.

**Comment by zadam** [comment]
Hi, Trilium needs SQLite and node.js - it can't function as a pure PWA. To do that, basically all backend code would have to be rewritten so that's quite unlikely to happen.
